### DISCUSSION TITLE
ScriptReference/Graphics.DrawMeshInstancedIndirect

### DISCUSSION THREAD

**Discussion by docscomments** [Developer-Notes-Extension/Unity-Notes#242] [Notes]
# ScriptReference/Graphics.DrawMeshInstancedIndirect

https://docs.unity3d.com/ScriptReference/Graphics.DrawMeshInstancedIndirect.html

**Comment by mvi** [comment]
Note you may need to supply this pragma to ensure that `PROCEDURAL_INSTANCING_ON` is defined (which in turn defines `UNITY_PROCEDURAL_INSTANCING_ENABLED` and `UNITY_ANY_INSTANCING_ENABLED`)  
`#pragma instancing_options procedural:setup`  
where setup is at the very least an empty function:  
`void setup() { }`  
See the "#pragma instancing_options" section of [GPUInstancing](https://docs.unity3d.com/Manual/GPUInstancing.html) for more details, specifically:  
  
> procedural:FunctionNameUse this to instruct Unity to generate an additional variant for use with Graphics.DrawMeshInstancedIndirect.  
At the beginning of the vertex Shader stage, Unity calls the function specified after the colon. To set up the instance data manually, add per-instance data to this function in the same way you would normally add per-instance data to a Shader. Unity also calls this function at the beginning of a fragment Shader if any of the fetched instance properties are included in the fragment Shader.

**Comment by docscomments** [comment]
**Comment by `@OswaldHurlem`:**  

This performs frustum culling based on the bounds parameter.

**Comment by radiatoryang** [comment]
Normals on mesh instances will be messed up, because unity_WorldToObject inverse is incorrectly calculated. Calculate the inverse matrix manually in the shader. (from https://forum.unity.com/threads/trying-to-rotate-instances-with-drawmeshinstancedindirect-shader-but-the-normals-get-messed-up.707600/)

```
float4x4 inverse(float4x4 input)
{
#define minor(a,b,c) determinant(float3x3(input.a, input.b, input.c))
    float4x4 cofactors = float4x4(
        minor(_22_23_24, _32_33_34, _42_43_44),
        -minor(_21_23_24, _31_33_34, _41_43_44),
        minor(_21_22_24, _31_32_34, _41_42_44),
        -minor(_21_22_23, _31_32_33, _41_42_43),
        -minor(_12_13_14, _32_33_34, _42_43_44),
        minor(_11_13_14, _31_33_34, _41_43_44),
        -minor(_11_12_14, _31_32_34, _41_42_44),
        minor(_11_12_13, _31_32_33, _41_42_43),
        minor(_12_13_14, _22_23_24, _42_43_44),
        -minor(_11_13_14, _21_23_24, _41_43_44),
        minor(_11_12_14, _21_22_24, _41_42_44),
        -minor(_11_12_13, _21_22_23, _41_42_43),
        -minor(_12_13_14, _22_23_24, _32_33_34),
        minor(_11_13_14, _21_23_24, _31_33_34),
        -minor(_11_12_14, _21_22_24, _31_32_34),
        minor(_11_12_13, _21_22_23, _31_32_33)
        );
#undef minor
    return transpose(cofactors) / determinant(input);
}
```

and then in your setup function: `unity_WorldToObject = inverse(unity_ObjectToWorld);`

**Comment by radiatoryang** [comment]
for culling individual instances, you have two (practical) options:
- cull via Job System... make sure it's Burst Jobs, it'll be way too slow otherwise https://ennogames.com/blog/frustum-culling-with-unity-jobs
- cull via Compute Shader... most of the code samples I found are way too complicated and (imo) over-engineered. Simplest understandable compute culling example I found was this: https://github.com/ColinLeung-NiloCat/UnityURP-MobileDrawMeshInstancedIndirectExample/blob/master/Assets/URPMobileGrassInstancedIndirectDemo/InstancedIndirectGrass/Core/CullingCompute.compute
